shard/src/ext/record_cog.py

```diff
@@ -0,0 +1,254 @@
+
+        # Header values that can't be hardcoded
+        data_chunk_size = self.num_channels * len(wav_data[0])
+        chunk_size = pack("<L", 36 + data_chunk_size)
+        data_chunk_size = pack("<L", data_chunk_size)
+        byte_rate = pack("<L", 48000 * self.num_channels * 2)
+        block_align = pack("<H", self.num_channels * 2)
+        n_chan = pack("<H", self.num_channels)
+
+        # RIFF chunk descriptor
+        self.f.write(b"\x52\x49\x46\x46")              # "RIFF", specifies RIFF file type
+        self.f.write(chunk_size)                       # Size of the file minus this and "RIFF"
+        self.f.write(b"\x57\x41\x56\x45")              # "WAVE", specifies wave subtype
+
+        # fmt sub chunk
+        self.f.write(b"\x66\x6d\x74\x20")              # "fmt ", starts format section
+        self.f.write(b"\x10\x00\x00\x00")              # 16, size of this part of header
+        self.f.write(b"\x01\x00")                      # 1, PCM mode
+        self.f.write(n_chan)                           # number of channels
+        self.f.write(b"\x80\xBB\x00\x00")              # 48000, sample rate of file
+        self.f.write(byte_rate)                        # byte rate
+        self.f.write(block_align)                      # number of bytes in an entire sample of all channels
+        self.f.write(b"\x10\x00")                      # Bits in a sample of one channel
+
+        # data chunk
+        self.f.write(b"\x64\x61\x74\x61")              # "data", in data header now
+        self.f.write(data_chunk_size)                  # size of the data chunk
+
+        # write the actual PCM data
+        for j in range(0, len(wav_data[0]), 2):
+            for i in range(self.num_channels):
+                self.f.write(wav_data[i][j:j + 2])     # make sure to write two bytes as sample size is 16 bits
+
+        self.event.set()
+
+
+class RecordCog(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+        self.recording = False
+        self.wav_file = None
+        self.event = AsyncThreadEvent()
+
+        # Ensure the opus library is loaded as it is needed to do voice things.
+        if not discord.opus.is_loaded():
+            discord.opus.load_opus("res/libopus.so")
+
+    @commands.command(alias=['record'])
+    async def start_recording(self, ctx):
+        """
+        Records the voice channel that the commanding user is currently in.
+        """
+        if self.recording:
+            await ctx.send("I'm already recording. Stop the current recording then start a new one.")
+            return
+
+        try:
+            await self.ensure_voice(ctx)
+        except commands.CommandError:
+            return
+
+        self.recording = True
+
+        members = [m for m in ctx.author.voice.channel.members if not m.bot]
+        self.wav_file = BytesIO()
+        sink = WavFile(self.wav_file, members, self.event, self.bot.user)
+        ctx.voice_client.listen(sink)
```

Comment:
is there a max time on this?

Reply:
Not currently. If you have a max file size I can estimate a max time to set.